Remove Liveness and add a Low Power mode to the Wallet

philipr-za opened the pull request:
## Description
In order to help reduce battery consumption this PR removes the Liveness service from the Wallet which is no longer required due to updates to the connectivity manager. 

This PR also adds two FFI functions to set Low power and Normal power modes. Currently these modes will change the Timeouts used by the two  base node monitoring protocols to poll the base node from 30 seconds to 5 minutes. The mobile client can switch to the low power mode when the app moves to the background when it doesn’t need to be as responsive and it will generate 10 times less network traffic and thus use less battery.

Closes https://github.com/tari-project/tari/issues/1963

@kukabi

## How Has This Been Tested?
Tests have been updated to exercise the new functions

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

kukabi commented:
Amazing, thanks!

sdbondi commented on base_layer/wallet/src/transaction_service/protocols/transaction_chain_monitoring_protocol.rs:
```diff
@@ -57,6 +57,7 @@ const LOG_TARGET: &str = "wallet::transaction_service::protocols::chain_monitori
 
 /// This protocol defines the process of monitoring a mempool and base node to detect when a Broadcast transaction is
 /// Mined or leaves the mempool in which case it should be cancelled
+#[allow(clippy::too_many_arguments)]
```
Should be on the `new` function

sdbondi commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -4014,6 +4014,69 @@ pub unsafe extern "C" fn wallet_coin_split(
+        Err(e) => {
+            error = LibWalletError::from(WalletError::TransactionServiceError(e)).code;
+            ptr::swap(error_out, &mut error as *mut c_int);
+            ()
```
📎 
```suggestion
    if let Err(e) = (*wallet)
        .runtime
        .block_on((*wallet).transaction_service.set_normal_power_mode()) {
            error = LibWalletError::from(WalletError::TransactionServiceError(e)).code;
            ptr::swap(error_out, &mut error as *mut c_int);
        }
```

Same for the function above

sdbondi commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -952,20 +965,27 @@ where
+                let (timeout_update_sender, timeout_update_receiver) = mpsc::channel(100);
                 self.mempool_response_senders.insert(tx_id, mempool_response_sender);
                 self.base_node_response_senders.insert(tx_id, base_node_response_sender);
+                self.timeout_update_senders.insert(tx_id, timeout_update_sender);
```
Wouldn't a single timeout event for the txn system work and be more efficient, rather than per transaction? Would, of course, need a broadcast channel to produce multiple receivers. 

One important note, with a broadcast channel the underlying Vec is preallocated to the given buffer size. 100 seems too large for how often the power mode will be changed - 5 or 10 should be plenty (`std::mem::size_of::<Duration>() == 16`(1600 bytes vs 160 bytes)

philipr-za replied:
Ja, I agree its neater. Was just in this mindset for the other channels.

sdbondi commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -1512,3 +1560,9 @@ where TBackend: TransactionBackend + Clone + 'static
+
+#[derive(Clone, Copy)]
+enum PowerMode {
+    LowPower,
```
nit: `Power` is redundant as context on variants is given by the enum type - `PowerMode::Low` and `PowerMode::Normal`

sdbondi reviewed (commented):
Looks good - I think a single broadcast channel fits better than an mpsc channel per transaction

philipr-za commented:
@StriderDM This PR adds two FFI functions that will need to be added to the JNI and Swift wrappers once it goes in.

philipr-za committed:
Remove Liveness and add a Low Power mode to the Wallet

In order to help reduce battery consumption this PR removes the Liveness service from the Wallet which is no longer required due to updates to the connectivity manager. 

This PR also adds two FFI functions to set Low power and Normal power modes. Currently these modes will change the Timeouts used by the two  base node monitoring protocols to poll the base node from 30 seconds to 5 minutes. The mobile client can switch to the low power mode when the app moves to the background when it doesn’t need to be as responsive and it will generate 10 times less network traffic and thus use less battery.

sdbondi commented on base_layer/wallet/src/transaction_service/config.rs:
```diff
@@ -24,19 +24,20 @@ use std::time::Duration;
+    pub base_node_monitoring_timeout: Duration,
     pub direct_send_timeout: Duration,
     pub broadcast_send_timeout: Duration,
+    pub low_power_polling_timeout: Duration, /* This is the timeout period that will be used when the wallet is in
+                                              * low_power mode */
```
nit:
```suggestion
    /// The timeout period that will be used when the wallet is in low_power mode
    pub low_power_polling_timeout: Duration, 
```

sdbondi reviewed (approved):
Looks good - let's see how much of a difference it makes
